### ISSUE TITLE
Presenting Binder at NIPS 2018

### ISSUE THREAD

**Issue by jzf2101** [jupyterhub/team-compass#52]
Hi Everyone,

I'm looking at NIPS 2018 and there are a number of workshops that are of particular note to the JupyterHub Community:

https://2018.mloss.org/cfp/
- I think a paper about binder sustainability would be of particular interest to this community
- or something about how people are using binder

and 

https://nips.cc/Conferences/2018/Schedule?showEvent=10919
- perhaps littlest jupyterhub would be something to submit here

additionally I think it'd be interesting to submit binder to the NIPS demo track:

https://nips.cc/Conferences/2018/CallForDemonstrations

Would anyone else be interested in attending? @betatim @minrk @yuvipanda @choldgraf @consideRatio ?

**Comment by choldgraf** [comment]
I am going to try and take a break from travel for a little bit, but I think these sound like appropriate tracks for sure. I can help folks prepare materials/proposals/etc if somebody wants to take a lead on this

**Comment by yuvipanda** [comment]
Thank you for taking the lead on this, @jzf2101. Unfortunately I won't have time to help much. Good luck!

**Comment by minrk** [comment]
I'm not going t be able to travel for the rest of 2018, but I'd be happy to help support developing materials

**Comment by jzf2101** [comment]
Updates:

We have a submission on MLOSS 

https://openreview.net/forum?id=BJepbQkJ5Q

@minrk if you could please provide feedback on the manuscript. 


Second, we have a Binder Demo for NIPS 

@cmd-ntrf has offered to help present.  We need to figure out who else may be able to come with us.  

@fperez has suggested that we create a separate deployment.

The first thing I can think of we can do is pull the fork from https://github.com/crowdAI/repo2docker so we can use nvidia docker on a deployment.

@rgiordan we'd appreciate your thoughts on what to put on the demo

**Comment by jzf2101** [comment]
I don't think we have to pull it to master- we can pull it to a branch.  I will turn this into a different format that's easier to process.

**Comment by rgiordan** [comment]
Was there a survey you wanted me to fill out, or would you like me to just
look through it and send free-form suggestions?

**Comment by choldgraf** [comment]
@rgiordan anything that you think would be useful to demonstrate in a conference like this, from your perspective, would be helpful to us :-)

**Comment by jzf2101** [comment]
@betatim 

>What is the thinking behind having a second (temporary?) deployment? I assume it would be too hard for us to maintain to deployments in the long term, especially if they are different. The clear advantage of a temporary deployment is that we can take shortcuts and there are no promises of long term support. 

The motivation is that our current deployment provides only 1 CPU node per user.  To reproduce a machine learning pipeline, you would likely need more compute, plus GPUs.  I think providing this amount of compute long term would be costly.  @fperez please chime in if you like.


>I think it would be worth discussing how we can do this so we can merge a lot of the work for the NIPS demo into mainstream binder and tooling. For example the crowdAI fork of repo2docker seems to be mostly a change in base image, we should have the discussion on to add that to r2d proper.

I was thinking of making this at least a branch for now so we can show how repo2docker works with nvidia docker.

**Comment by fperez** [comment]
What I thought, in chatting with @jzf2101, was that we could spin up *just for a few days* (around the conference) a few more expensive nodes to demo these capabilities. We wouldn't keep them up forever, but it would be a good demonstration of how one can use Binder to run more complex workloads that may require custom hardware (obviously having to pay for it).

We would shut it down a few days after the conference, but having the image ready to go, with docs, would also help others who may be interested in having their own custom deployments with these features.

Just a thought, not married to it...

**Comment by choldgraf** [comment]
I think it's a good idea and lets us highlight the idea that "mybinder.org is not the only binderhub in existence, nor should it be"

**Comment by jzf2101** [comment]
Yeah, it's something we have explicitly said in the MLOSS workshop submission

**Comment by betatim** [comment]
For all the below apply the "we gotta be realistic about how much resources we have to pull this off by NIPS" caveat emptor ;)

Setting up a second BinderHub with more resources per repo that is only live for the duration of the conferences requires the least amount of resources. My thought was "if we spend a little extra do we get lasting improvements from this?". Two examples:

* we use a hacked repo2docker that uses a different base image. Easy to do, will take some debugging and effort to get example repos that work with this hacked version. Having a feature that lets us deal with different base images or adding GPU support is something we want in repo2docker proper. Is it possible to spend a bit extra in effort to get it merged/nearly merged into repo2docker instead of making a fork?

* more resources to a binder. Being able to give different repos different amounts of resources would be a cool feature to have for people hosting a BinderHub at their institution and mybinder.org. Instead of setting up a vanilla BinderHub which gives more resources to each repo we could work on adding differential resource allocation to BinderHub and use mybinder.org. This would give us a lasting benefit from doing the NIPS deployment.

I don't think that effort is wasted by setting up a temporary BinderHub with hacked tools. Instead I see it as "if we spend a bit extra we get longer lasting change". I don't know if we can afford to spend a little extra or not. I do think we should discuss it (and then decide that in an ideal world we'd do this but today pragmatism has to win).

I know of myself that taking the "easy, short term option" is super attractive but that often if someone makes me explain what I am doing for even a second I realise that the long term option isn't as much (or no) extra work.

---

If it is worth doing, it is worth doing properly! -- someone in a world of spherical cows and no friction

**Comment by jzf2101** [comment]
I don't disagree, I think we can use this as an opportunity to make improvements to the project.

**Comment by jzf2101** [comment]
@mpacer you've attended the conference in the past- would you be interested in helping with this issue and presenting?

**Comment by jzf2101** [comment]
FYI I wanted to share the Workshop Review with you all:

> Reproducibility is a hard issue in data science and machine learning research. Binder offers a very practical approach, centered on capturing the environment of the scientist doing the work, and by hosting the result in a web notebook.
> 
> This approach is attractive, as it reduces the burden of making one's experiments available to other researchers. It also doesn't reinvent any of the necessary technology, as it is layered on top of established OSS packages such as Docker and Jupyter, and it re-uses established infrastructure such as GitHub.
> 
> I believe this submission is valuable to the NIPS audience, and should be accepted as a demo.
> 
> In that demo, it might be helpful for the authors to provide guidance on a couple of questions I have:
> 
>   * Is it possible to manage multi-repository projects? Where one needs e.g. a patched version of NumPy to reproduce the results?
>   * How does one test whether the binder version of a repository produces the same results as the one run on the infrastructure of the scientist? Is there, e.g. a testing facility?
> 

https://openreview.net/forum?id=BJepbQkJ5Q

We should try to come up with a specific response for the reviewer.

**Comment by betatim** [comment]
> * Is it possible to manage multi-repository projects? Where one needs e.g. a patched version of NumPy to reproduce the results?

We can make a repo that pulls in a modified version of Numpy (via a `requirements.txt` line to a fork of Numpy). Has anyone ever tried to modify numpy code? Otherwise maybe we demo it with a modified version of scikit-learn where I know how to change things.

> * How does one test whether the binder version of a repository produces the same results as the one run on the infrastructure of the scientist? Is there, e.g. a testing facility?

I think this is harder as the answer is "hopefully the authors provided good notebooks/demos/scripts that produce the key results of the paper". Pretty hard to provide a generic tool or automation for this I think as what a result is can vary so much from paper to paper. Maybe there is something out there already where people can upload in machine-readable format "results"? Worth talking to https://www.openml.org/?

**Comment by jzf2101** [comment]
>How does one test whether the binder version of a repository produces the same results as the one run on the infrastructure of the scientist? Is there, e.g. a testing facility?

I think we should reference `start` and `postBuild` and discuss future work on `verify`.  

@ledell collaborates with the openml people so maybe she would have some ideas?

**Comment by ledell** [comment]
@jzf2101 @betatim 

> >  How does one test whether the binder version of a repository produces the same results as the one run on the infrastructure of the scientist? Is there, e.g. a testing facility?

> I think we should reference start and postBuild and discuss future work on verify.
> @ledell collaborates with the openml people so maybe she would have some ideas?

OpenML is meant to be used for ML experiments and in Python, the most straight-forward way to use it is via the [OpenML sklearn Python API](https://docs.openml.org/sklearn/).  So in the context of ML experiments, you could use OpenML for reproducibility, however for generic/non-ML workflows, it's probably not the most straight-forward use-case.

**Comment by Zsailer** [comment]
> * how does linking to the original experiment work, after it has been cloned and modified (eg fixed or extended)? automatic?

You can create Binders from different commits/tags/branches in a given Github repository. We could demonstrate a "best practice" for archiving iterations of a repo. Either create a tag or new branch when the notebooks are changed and create a Binder for that tag/branch.

> * does one see a history of changes to an experiment (setup, results, etc), differences to original experiment?

Then, add Binder links/badges to a list on the front page of the repository's README for each iteration. This is basically a change-log for the Binder repo.

It's not full proof. Someone could mess with the commit logs and remove history of results they don't want public. But, it does allow for some "history of changes".

**Comment by choldgraf** [comment]
I think we need to clarify how Binder fits in with the broader ecosystem. E.g., people asking "how can we keep track of the past results of an experiment" tells me that we aren't doing a good job of explaining how Binder fits into pre-existing versioning/discovery/sharing systems, it does not (not does it want to) replace them.

**Comment by betatim** [comment]
What needs doing in terms of deploying a BinderHub for this demo?

**Comment by jzf2101** [comment]
First we need to set up the deployment.  @choldgraf are you in charge of the project.

**Comment by choldgraf** [comment]
I don't believe so, and I'm probably not the right person to set up a deployment like this :-/  somebody more experienced with Kubernetes would be better

**Comment by betatim** [comment]
To get a deployment up and running we need to know which billing account this deployment should be connected to and what will be different compared to the mybinder.org deployment. Some potential differences:
* domain name
* per user resources
* theme
* repo restrictions/white listed repos only

**Comment by choldgraf** [comment]
I think we should:

* use the binder-prod billing account
* set up a new k8s cluster with fancier machines that run whatever code @jzf2101 imagines running
* Find a set of repositories to highlight (or prepare some of our own)
* Over-provision the cluster only for a few days before/after NIPS
* Pre-pull the images of any repos we want to highlight

**Comment by betatim** [comment]
👍  to all of Chris' points. I might turn over-provision into "run with a minimum number of nodes that has headroom built in but still tries to save some cash by not having all nodes running all the time".

Some cost estimates from the calculator (assuming the instance will be live for 72h). If per user we provide the equivalent of: n1-standard-2 (2vCPU, 7.5GB RAM), 1 NVIDIA TESLA K80 to each user/pod that will cost $40.

If this instance were to be as popular as mybinder.org (400 concurrent pods/users) we'd spend $16k in those three days. Maybe 100 concurrent users is more realistic(?) which comes in at $4k or roughly a month of mybinder.org. Conclusion: It seems like we can afford this.

Has someone setup a BinderHub with GPUs and built a repo that works? If not the step towards that is inspecting the nvidia-docker image to work out which extra packages need installing and making a repo that works with the current repo2docker or making a patched version of repo2docker that uses the nvidia base image instead of ubuntu. I have a feeling this requires someone with access to a GPU to test this. Who do we know that has this?

**Comment by choldgraf** [comment]
cc @jhamman maybe he's played around with fancy shmancy binderhub deployments on GCP?

**Comment by jhamman** [comment]
Thanks for the ping here @choldgraf. We haven't gone down the GPU path in earnest yet but its certainly in scope for binder.pangeo.io. If someone wants to play around with this using our deployment, I'd be happy to work with them. Just to describe what we do currently.

We have three node pools:
1. core pool - this runs binderhub, jupyterhub, etc. No notebooks or dask workers here.
2. user pool - just builder and jupyter pods. We typically limit users to 2 cpu and 8ish gb RAM.
3. worker pool - just dask workers. Users can get a bunch of these though we try to limit it to something like 50 pods per user.

If there were an easy way to tell binderhub to put certain notebooks in certain node pools, that would be a nice way to start working on this problem of custom resources. In the short term, we could build a node pool with some GPUs in it and give things a try.

**Comment by betatim** [comment]
@jzf2101 do you have access to a computer with a GPU? If not how can we make that happen?

How many days before NIPS do we want this to be ready and how do we put the cluster into a parking position with 0 GPUs. @consideRatio mentioned that auto-scaling seems to be a bit flaky with GPUs right now. I am thinking we should make two node pools: one "core pool" which can be small and idle away and a second one with GPU instances that we manually scale to zero when not working on the cluster or dmeo'ing it.

From https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/994#issue-373992464 it looks like we won't have to modify repo2docker to get started as everything is a conda/apt package. This is great news. One thing that isn't clear to me is if we can test docker images built that way on a local machine that has a GPU or if the setup described in that post relies on some crazy integration provided by GCP? @jzf2101 do you have access to a GPU locally? This would allow us to get started on making some of the example repos work without blocking on setting up the cluster which can happen in parallel.

**Comment by jzf2101** [comment]
I can set up a local VM if that works for you @betatim ?

**Comment by choldgraf** [comment]
@jzf2101 I just added you to the Binder GCP account as an owner. Friendly reminder to anybody else that you should ask if you'd like to be added to the account! I think anybody can add anybody else since you're all listed as "owners"

**Comment by jzf2101** [comment]
@jhamman could we also feature your pangeo deployment? If you have any binders connected to an academic publication it'd be good to know?

**Comment by jhamman** [comment]
@jzf2101 - please feel free to feature the pangeo deployment. Just let us know if there are any specific things to coordinate on. 

I'm not sure on the academic pubs just yet. The Pangeo concept/workflow is quite new so I would bet we don't have any pure Pangeo binders just yet. We have lots of tutorials/examples though that we can share.

**Comment by jzf2101** [comment]
You may end up getting additional traffic the first week in December fyi @jhamman

**Comment by jzf2101** [comment]
Can we set up a time to try to put out a deployment? Because of earlier deadlines I have, I'd like to have a blank binder up a week by the 21st and get binder images ready by the end of that week the 25th?

**Comment by betatim** [comment]
> I can set up a local VM if that works for you @betatim ?

I won't have time (and don't have access to a GPU) to help make repos run beyond pointing to examples and linking documentation/how to do things in repo2docker. So what ever works for you to get access to a GPU.

The next step (I think) is to setup a new BinderHub deployment with a cheap instance type (no GPUs), get it debugged and make a repository to share the configuration and secrets. I could work on that this weekend.

One thing that isn't documented in the Zero2BinderHub guide is how to get HTTPS setup, Maybe @yuvipanda can write a short thing on that?

Otherwise it looks like it should be a matter of following the instructions and cribbing from mybinder.org-deploy.

@choldgraf are you the person who has access to the mybinder.org DNS? We will need to setup the cluster first before we know the IP address to configure but it would be good to know who to talk to about it.

**Comment by jzf2101** [comment]
One thing we may be aware of though is

https://github.com/jupyter/docker-stacks/issues/745#issuecomment-436284752

**Comment by jzf2101** [comment]
So if we divide and conquer- I can pick some repos and make sure they run with GPUs?

If we get that done would you be able to add some K80s to the spawner?

**Comment by choldgraf** [comment]
I think I do have DNS access, happy to work with folks to change things (though I believe that I'm using a lastpass credential share that @Carreau sent me so I don't actually know the passwords etc)

**Comment by jzf2101** [comment]
@betatim I can start with a simple JupyterHub?

**Comment by jzf2101** [comment]
FYI @yuvipanda mentioned that we may not need HTTPS

**Comment by jzf2101** [comment]
@aaronculich said he'd be able to help

**Comment by betatim** [comment]
@jzf2101 if you know how to and have the time to do so please go ahead with setting up a JupyterHub on GKE and setting up a repo for it. Then we can layer a BinderHub on top.

> I can also try to cobble together an r2d branch that can make them run?

From reading https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/994#issue-373992464 I don't think we need to modify repo2docker to make things work.

**Comment by choldgraf** [comment]
So it sounds like the order of operations here is:

Tech stuff
1. Choose a Docker image with the environment we want (or as @jzf2101 suggests, build a docker image with the cuda version of repo2docker)
2. Set up a JupyterHub that can run CUDA code
3. If that works, then set up a BinderHub wrapped around this JupyterHub

Content stuff
1. Find a few repositories of ML methods that we can highlight. Ideally papers, but anything that seems "interesting" should be fine IMO.
2. Anything else? Training / information on building your own Binder content?

**Comment by jzf2101** [comment]
Sure let's close and move this one somewhere else
